migration/dialopts.go

```diff
@@ -7,15 +7,29 @@ import (
 	"time"
 
 	"github.com/juju/juju/api"
+	"github.com/juju/juju/core/migration"
 )
 
+// NewLoginProvider returns a LoginProvider for the target controller
+// using the given TargetInfo. If TargetInfo contains a session token,
+// it creates a SessionTokenLoginProvider for authenticating with JIMM controllers.
+// If no session token is present, it returns nil, which is treated as legacy
+// authentication i.e. username/password or macaroons.
+func NewLoginProvider(targetInfo migration.TargetInfo) api.LoginProvider {
+	if targetInfo.Token != "" {
+		return api.NewSessionTokenLoginProvider(targetInfo.Token, nil, nil)
+	}
+	return nil
```

Comment:
I wish we didn't use nil rather than a concrete "legacy" provider

Reply:
Yeah me too, it's something that would require more work to take bits of logic out of the `api.Open` function.

Reply:
Just a few lines above there is an existing TODO,
>	// TODO (alesstimec, wallyworld): login provider should be constructed outside
	// of this function and always passed in as part of dial opts. Also Info
	// does not need to hold the authentication related data anymore. Until that
	// is refactored we fall back to using the user-pass login provider
	// with information from Info.